### PR TITLE
WIP UI schema

### LINKED ISSUE #1232: Busola can be extended to support custom views for custom CRDs
As [Sebastian](https://github.com/kyma-project/community/blob/main/manifesto/personas.md#sebastian---tech-lead) and [Mathew](https://github.com/kyma-project/community/blob/main/manifesto/personas.md#matthew---cloud-native-software-engineer) I would like to be able to extend Busola by configuration. I would expect that with a small configuration I can enable a custom UI in Busola where I can List, View Detail, and Edit and Delete a component

**Acceptance Criteria**

- [x] The component can have a configurable menu item that can be cluster level or namespace level
- [x] The component can have a configurable list where columns can be picked up and have at least 3 components
- [x] The Component can have configurable detail page and we will provide at least 3 components and we can choose 2 layout types ( stacked and 2 columns )
- [x] The Component can have a configurable Create Page with Simple and Advanced sections
- [x] The Component can support multiple translations and multiple languages
- [ ] The configuration is versioned
- [ ] The configuration is clearly documented with examples - think about using json schema to autogenerate (https://json-schema.org/implementations.html#documentation-generators)
- [x] This will be feature flagged and disabled by default
- [ ] Follow Project DOD

*Useful Links**

- Inspiration for the structure - https://jasonette.com/
- Inspiration for the structure - https://ui-schema.bemit.codes/
- https://www.storyblok.com/tp/react-dynamic-component-from-json
- [SPIKE](https://github.com/kyma-project/busola/issues/1222)

### PR DESCRIPTION
**Pull request by mrCherry97**
**Description**

Changes proposed in this pull request:

- ...
- ...
- ...

**Related issue(s)**
Closes #1232

### CODE REVIEW

File: core-ui/src/shared/components/ResourceDetails/ResourceDetails.js [hunk 1]
```diff
@@ -51,7 +51,7 @@ ResourceDetails.propTypes = {
 
 ResourceDetails.defaultProps = {
   customColumns: [],
-  customComponents: [],
+  customComponentscustomComponents: [],
```
Comment (Wawrzyn321) [inline core-ui/src/shared/components/ResourceDetails/ResourceDetails.js:54]:
```suggestion
  customComponents: [],
```

File: core-ui/src/components/Extensibility/ExtensibilityList.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    );
+  }
+  listProps.createFormProps = { resource: resMetaData };
+  listProps.resourceName = t('labels.name', {
+    defaultValue: resMetaData.navigation.label,
+  });
+  listProps.description = t('labels.description', {
+    defaultValue: '',
+  });
+  listProps.customColumns = (resMetaData.list || []).map(column => ({
```
Comment (Wawrzyn321) [inline core-ui/src/components/Extensibility/ExtensibilityList.js:34]:
```suggestion
  listProps.customColumns = (resMetaData.list?.columns || []).map(column => ({
```

File: core-ui/src/components/Extensibility/ExtensibilityDetails.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  const breadcrumbs = [
+    {
+      name: resMetaData.navigation.label || resMetaData.navigation.path,
+      path: '/',
+      fromContext: resMetaData.navigation.path,
+    },
+    { name: '' },
+  ];
+  return (
+    <TranslationBundleContext.Provider value={resMetaData.navigation.path}>
```
Comment (Wawrzyn321) [inline core-ui/src/components/Extensibility/ExtensibilityDetails.js:39]: I guess we can remove it from this component, as it's already used in `ExtensibilityDetails` below.

File: core-ui/src/components/Extensibility/components/Columns.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import React from 'react';
+
+import { Widget } from './Widget';
+
+export function Columns({ value, structure, schema }) {
+  return (
+    <div class="panel-grid">
```
Comment (Wawrzyn321) [inline core-ui/src/components/Extensibility/components/Columns.js:7]:
```suggestion
    <div className="panel-grid">
```

**Review by Wawrzyn321 — CHANGES_REQUESTED** [review]
Ohhh I forgot I need to submit my review, so you can see my comments 🤦

**Comment by Wawrzyn321** [comment]
Small change - _createNavigationNodes_ in `core/src/luigi-config/feature-discovery.js:124` requires 4th argument, `customResources`.

File: core-ui/src/components/Extensibility/ExtensibilityList.js (hunk 2 above)
Comment (qbalukom) [reply core-ui/src/components/Extensibility/ExtensibilityList.js:34]: `resMetaData.list` is actually the correct behaviour - not all specs are currently updated, though

File: core-ui/src/components/Extensibility/ExtensibilityDetails.js (hunk 3 above)
Comment (qbalukom) [reply core-ui/src/components/Extensibility/ExtensibilityDetails.js:39]: yes, absolutely. good catch. even more, that's what the extra component was created for :)

File: core/src/assets/customResources/customResources2Schema.json [hunk 5]
```diff
@@ -0,0 +1,354 @@
+{
```
Comment (wchrapka) [inline core/src/assets/customResources/customResources2Schema.json:1]: Is this needed? This looks like my old file that can be deleted

File: core-ui/src/resources/helpers.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  const routerParams = useParams();
+  const queryParams = new URLSearchParams(window.location.search);
+  const { i18n, t } = useTranslation();
+  const resourceUrl = getResourceUrl();
+  return {
+    hasDetailsView: queryParams.get('hasDetailsView') === 'true',
+    readOnly: queryParams.get('readOnly') === 'true',
+    resourceUrl,
+    resourceType: resourceType,
+    resourceName: t(resourceI18Key || '') ? t(resourceI18Key) : resourceI18Key,
```
Comment (wchrapka) [inline core-ui/src/resources/helpers.js:17]: `t(resourceI18Key)` will return `resourceI18Key` by default if no transaltion
```suggestion
    resourceName: t(resourceI18Key || ''),
```

File: core-ui/src/components/Extensibility/ExtensibilityCreate.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+      pluralKind={resourceType}
+      singularName={resourceType}
+      resource={resource}
+      setResource={setResource}
+      formElementRef={formElementRef}
+      createUrl={resourceUrl}
+      setCustomValid={setCustomValid}
+      onlyYaml={!schema}
+    >
+      <ResourceSchema
```
Comment (wchrapka) [inline core-ui/src/components/Extensibility/ExtensibilityCreate.js:37]: this one shouldn't be displayed only in the "create" mode?

File: core-ui/src/components/Extensibility/ExtensibilityDetails.js [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+      {...detailsProps}
+    />
+  );
+};
+
+export const ExtensibilityDetails = () => {
+  const resMetaData = useGetCRbyPath();
+
+  return (
+    <TranslationBundleContext.Provider value={resMetaData.navigation.path}>
```
Comment (wchrapka) [inline core-ui/src/components/Extensibility/ExtensibilityDetails.js:64]: Could this be wrapped in an ErrorBoundary (custom or `core-ui/src/shared/components/ErrorBoundary/ErrorBoundary.js`)

File: core-ui/src/components/Extensibility/ExtensibilityList.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  }));
+  return (
+    <ResourcesList createResourceForm={ExtensibilityCreate} {...listProps} />
+  );
+};
+
+export const ExtensibilityList = () => {
+  const resMetaData = useGetCRbyPath();
+
+  return (
```
Comment (wchrapka) [inline core-ui/src/components/Extensibility/ExtensibilityList.js:48]: Could this be wrapped in an ErrorBoundary ?

File: Dockerfile [hunk 10]
```diff
@@ -19,8 +19,8 @@ RUN make resolve
 RUN make validate
 RUN make pull-licenses
 
-RUN cd /app/core && make test && make build
-RUN cd /app/core-ui && make test && make build
```
Comment (wchrapka) [inline Dockerfile:23]: reminder, let's not forget about this

File: core-ui/src/components/Extensibility/ResourceSchema.js [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+import { METADATA_SCHEMA } from './metadataSchema';
+
+const [firstPlugin, ...otherPlugins] = formWidgets.pluginStack;
+const widgets = {
+  ...formWidgets,
+  pluginStack: [firstPlugin, SchemaRulesInjector, ...otherPlugins],
+};
+
+function FormContainer({ children }) {
+  return <div container>{children}</div>;
```
Comment (wchrapka) [inline core-ui/src/components/Extensibility/ResourceSchema.js:26]: That container probably does nothing

File: core/src/assets/customResources/customResources2Schema.json (hunk 5 above)
Comment (qbalukom) [reply core/src/assets/customResources/customResources2Schema.json:1]: indeed

File: core-ui/src/components/Extensibility/ResourceSchema.js (hunk 11 above)
Comment (qbalukom) [reply core-ui/src/components/Extensibility/ResourceSchema.js:26]: yeah, but ui-schema requires a wrapper component

Comment (qbalukom) [reply core-ui/src/components/Extensibility/ResourceSchema.js:26]: (the `container` flag is the key here)

File: core-ui/src/components/Extensibility/ExtensibilityCreate.js (hunk 7 above)
Comment (qbalukom) [reply core-ui/src/components/Extensibility/ExtensibilityCreate.js:37]: We only display the form in create mode for now. We can add the distinction when we implement edit mode for extensibility

File: core-ui/src/components/Extensibility/ExtensibilityDetails.js (hunk 8 above)
Comment (qbalukom) [reply core-ui/src/components/Extensibility/ExtensibilityDetails.js:64]: good idea

File: Dockerfile (hunk 10 above)
Comment (qbalukom) [reply Dockerfile]: done

File: core/src/i18n/en.yaml [hunk 12]
```diff
@@ -977,6 +977,8 @@ events:
     source: Source
     type: Type
   title: Events
+extensibility:
+  error: Something went wrong. Please confirm the extension configuration is correct.
```
Comment (grego952) [inline core/src/i18n/en.yaml:981]:
```suggestion
  error: Your extension configuration is incorrect.
```
Taking into account our [UI guidelines}(https://github.com/kyma-project/community/blob/main/docs/guidelines/content-guidelines/04-ui-elements.md#error-messages) I'd go with something like this + maybe we can add a link that shows the user how to fix the configuration?

**Review by grego952 — APPROVED** [review]

**Review by Wawrzyn321 — APPROVED** [review]

**Pull request merged**